Cart

geekarist opened the pull request:
J'ai ajouté la page Panier, on y accède en ajoutant un article, on peut supprimer ce qui a été ajouté ou retourner à la page Catalogue.

Il y a encore un problème de consommation de ressources, au bout d'un certain nombre d'aller-retours entre `BookListActivity` et `CartActivity` il y a des `OutOfMemoryError`, et j'ai souvent des messages "application may be doing too much work on its main thread". Pour l'instant je ne sais pas trop d'où ça vient.

geekarist committed:
Create CartActivity

Commit:
Add listener on continueShopping Button

Commit:
Change layout & labels

geekarist committed:
Delete unused variable

geekarist committed:
Finish CartActivity on continueShoppingButton click

geekarist committed:
Create CartAdapter and use it in CartActivity

geekarist committed:
Fix exceptions when adding an item to the Cart

Commit:
Add image to cart layout

geekarist committed:
WIP: implement CursorAdapter with SQLiteDatabase

Commit:
WIP: insert a Book using CartAdapter

geekarist committed:
WIP: CursorAdapter view methods and insert

- Implement `CursorAdapter.newView` and `bindView`
- Handle transactions in `insert`
- Get book from Cursor

Commit:
Update Cursor in the CartAdapter

Commit:
Move CartAdapter to Application singleton

Commit:
Move Cursor creation to CartDatabaseHelper

geekarist committed:
Use the LoaderManager to create the Cursor

 It was created on the UI thread.

Commit:
Clear Adapter on BookListActivity and CartActivity stop

There were OutOfMemoryErrors when navigating between BookListActivity
and CartActivity. It was due to the Adapters being recreated on start of
the Activities, without ever being cleared.

Commit:
Improve CartActivity layout

geekarist committed:
Implement removeFromCartButton

Commit:
Display Cart total

- In CartDatabaseHelper: new method for computing the total

- In CartActivity: fill mTotalView with the total in
LoaderCallbacks.onLoadFinished

geekarist committed:
Update total when the Cart is changed

geekarist committed:
Move loader init to `onCreate()`

Commit:
Fix OutOfMemoryErrors

The book cover bitmaps had to be resized to a height of 200 px, while
keeping the aspect ratio. See https://goo.gl/WeMhJg.

Commit:
Rename BookListActivity and BookCatalogAdapter

geekarist committed:
Create unit test for CartActivity

Commit:
Extract CartLoaderCallbacks

geekarist commented:
Problème de ressources corrigé en redimensionnant les images avant de les afficher.

Commit:
Merge branch 'master' into cart

# Conflicts:
#	app/src/main/java/com/github/geekarist/henripotier/CatalogActivity.java

jeremiemartinez commented on app/src/main/java/com/github/geekarist/henripotier/Book.java:
```diff
@@ -1,12 +1,16 @@
-public final class Book {
+import java.io.Serializable;
+
+public final class Book implements Serializable {
```
Serializable n'est pas efficace en Android. Il faut préférer l'utilisation de Parcelable.

jeremiemartinez commented on app/src/main/java/com/github/geekarist/henripotier/CartActivity.java:
```diff
@@ -0,0 +1,76 @@
+            }
+        });
+
+        Book purchasedBook = (Book) getIntent().getSerializableExtra("purchasedBook");
```
Préféré Parcelable.

jeremiemartinez commented on app/src/main/java/com/github/geekarist/henripotier/CartActivity.java:
```diff
@@ -0,0 +1,76 @@
+    protected void onStart() {
+        super.onStart();
+
+        continueShoppingButton.setOnClickListener(new View.OnClickListener() {
```
@OnClick de ButterKnife ?
Sinon pourquoi dans le onStart.

jeremiemartinez commented on app/src/main/java/com/github/geekarist/henripotier/CartAdapter.java:
```diff
@@ -0,0 +1,77 @@
+
+    @Override
+    public View newView(Context context, Cursor cursor, ViewGroup parent) {
+        LayoutInflater inflater = LayoutInflater.from(mContext);
```
inline ?

jeremiemartinez commented on app/src/main/java/com/github/geekarist/henripotier/CartAdapter.java:
```diff
@@ -0,0 +1,77 @@
+    public void bindView(View view, Context context, Cursor cursor) {
+        ButterKnife.bind(this, view);
+        final Book book = mDbHelper.getBook(cursor);
+        mTitleView.setText(book.title);
```
Pourquoi ne pas faire une vue custom ? Dans newView tu la crée et dans bindView tu casts. Ca permet de garder les responsabilités bien séparé. CartViewItem par exemple.

jeremiemartinez commented on app/src/main/java/com/github/geekarist/henripotier/CartDatabaseHelper.java:
```diff
@@ -0,0 +1,69 @@
+
+import timber.log.Timber;
+
+class CartDatabaseHelper extends SQLiteOpenHelper {
```
alala les joies de SQLite en Android :) Y'a quelques ORM qui existent si cela t'interessent. ORMLite notamment. Si tu aimes la programmation fonctionnelle. Tu as SqlBrite il me semble.

geekarist replied:
Oui je vais regarder les ORM, mais je voulais expérimenter au moins une fois "les joies de SQLite" :)

jeremiemartinez commented on app/src/main/java/com/github/geekarist/henripotier/CartDatabaseHelper.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    public Book getBook(Cursor cursor) {
+        int id = cursor.getInt(0);
```
Conseil : Ecrire un TU pour ca.

geekarist replied:
Juste pour getBook ?
Ca me semble moins intéressant de tester le reste de la classe, c'est surtout un passe-plat pour les requêtes sql.

jeremiemartinez replied:
En fait. Tu t'apercevras que avoir un TU sur le mapping entre ton cursor et ton POJO, ca t'evite de belles galeres d'echange de champs quand tu feras evoluer ton modele.

jeremiemartinez commented on app/src/main/java/com/github/geekarist/henripotier/CartLoaderCallbacks.java:
```diff
@@ -0,0 +1,38 @@
+
+    @Override
+    public void onLoaderReset(Loader<Cursor> loader) {
+        // TODO
```
Pas sur, mais loadInBackground ?

jeremiemartinez commented on app/src/main/java/com/github/geekarist/henripotier/CatalogActivity.java:
```diff
@@ -57,6 +57,12 @@ public void failure(RetrofitError error) {
     @Override
+    protected void onStop() {
+        super.onStop();
+        mAdapter.clear();
```
Pourquoi ?

jeremiemartinez commented on app/src/main/java/com/github/geekarist/henripotier/CatalogAdapter.java:
```diff
@@ -71,17 +72,23 @@ public View getView(final int position, View convertView, ViewGroup parent) {
         mTitleView.setText(b.title);
         mPriceView.setText(parent.getResources().getString(R.string.price, b.price));
-        Picasso.with(mContext).load(b.cover).placeholder(R.drawable.book_cover_placeholder).into(mImageView);
+        Picasso.with(mContext).load(b.cover).resize(200, 200).centerInside()
```
Extraire une constante : private static final int SIZE = 200; ?

jeremiemartinez commented on app/src/main/java/com/github/geekarist/henripotier/CatalogAdapter.java:
```diff
@@ -71,17 +72,23 @@ public View getView(final int position, View convertView, ViewGroup parent) {
-                String msg = mContext.getResources().getString(R.string.buy_book_msg, b.title);
-                Toast.makeText(mContext, msg, Toast.LENGTH_LONG).show();
+                Intent intent = new Intent(mContext, CartActivity.class);
+                intent.putExtra("purchasedBook", b);
```
Extraire une constante : EXTRA_PURCHASED ? Comme ca quand tu feras le getStringExtra tu pourras la réutiliser.

jeremiemartinez commented on app/src/main/res/layout/activity_book_item.xml:
```diff
@@ -6,32 +6,34 @@
 
     <LinearLayout
-        android:layout_width="match_parent"
+        android:layout_width="fill_parent"
```
utilise match_parent. Derrière c'est la meme chose.

jeremiemartinez commented on app/src/main/res/layout/activity_book_item.xml:
```diff
@@ -6,32 +6,34 @@
-        android:layout_width="70dp"
-        android:layout_height="100dp" />
+        android:layout_width="90dp"
+        android:layout_height="fill_parent"
```
utilise match_parent. Derrière c'est la meme chose.

jeremiemartinez commented on app/src/main/res/layout/activity_cart_item.xml:
```diff
@@ -5,30 +5,37 @@
-        android:layout_height="100dp" />
+        android:id="@+id/cartItemImageView"
+        android:layout_width="90dp"
+        android:layout_height="fill_parent"
```
utilise match_parent. Derrière c'est la meme chose.

jeremiemartinez commented on app/src/main/res/layout/activity_cart_item.xml:
```diff
@@ -5,30 +5,37 @@
 
     <LinearLayout
-        android:layout_width="match_parent"
+        android:layout_width="fill_parent"
```
utilise match_parent. Derrière c'est la meme chose.

jeremiemartinez commented on app/src/main/res/layout/activity_cart_list.xml:
```diff
@@ -6,24 +6,16 @@
     <TextView
         android:id="@+id/total"
-        android:layout_width="match_parent"
+        android:layout_width="fill_parent"
```
utilise match_parent. Derrière c'est la meme chose.

jeremiemartinez commented:
Voila quelques commentaires. Hésites pas si c'est pas clair !

Commit:
Use @OnClick on continue_shopping button

Commit:
Inline inflater variable

Commit:
Extract BookView

geekarist committed:
Create layout file for BookView

geekarist committed:
Declare removeButton listener in CartAdapter

geekarist committed:
Use book view in CatalogAdapter

geekarist committed:
Fix exception on book selection

CartAdapter's book item layout did not contain an element for its
BookView. It was preventing ButterKnife to bind the BookView.

geekarist committed:
Parameterize choose button label in BookView

Commit:
Remove onStop in activities

geekarist committed:
Extract EXTRA_PURCHASED constant

geekarist committed:
Replace fill_parent with match_parent

geekarist committed:
Move ButterKnife.bind() to onFinishInflate()

Commit:
Improve cursor handling in CartActivity

- Add a releaseCursor() method which releases the cart cursor reference
- Separate CartAdapter creation and cursor swapping

geekarist commented on app/src/main/java/com/github/geekarist/henripotier/CartLoaderCallbacks.java:
```diff
@@ -0,0 +1,38 @@
+
+    @Override
+    public void onLoaderReset(Loader<Cursor> loader) {
+        mAdaptable.releaseCursor();
```
Apparemment onLoaderReset() est appelée quand le Loader s'apprête à libérer le Cursor.
Du coup c'est releaseCursor() qui va libérer la référence au Cursor en faisant un swapCursor(null).

jeremiemartinez replied:
Merci pour l'info. Ca fait longtemps :)

geekarist committed:
Unit test Cursor to Book mapping

Commit:
Delete unused Notifiable interface

Commit:
Implement Parcelable instead of Serializable
